### PR TITLE
deps,src: workaround vm.runInContext issue with top level "var"

### PR DESCRIPTION
**Pull request by MSLaguana**
addresses https://github.com/nodejs/node-chakracore/issues/420

ChakraCore doesn't handle global objects in the same way as v8, and
in particular ChakraCore doesn't support all proxy interceptors there.
We attempted to work around this via interceptors in the prototype
chain, but that was an imperfect solution, and a top-level "var"
declaration would end up adding a property to the global object in
a context without going through the interceptors.

To work around this in simple cases, we now take a snapshot of
properties on the global before running in a context, and diff the
changes at the end to patch up changes in the sandbox.

This is NOT a complete solution, and in particular will not work if
there is any sort of asynchronous operation that changes the state
of the global object. It also has behavioral differences if the
global object is writable but the sandbox object is not; in node-v8
this will result in the variable silently not being set in the
context and instead the value from the sandbox being used, while
here we discover the mismatch too late and the new value has been
set on the global.

##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

**Comment by MSLaguana** [comment]
CI: https://ci.nodejs.org/job/chakracore-test-pull-request/262/

**Comment by MSLaguana** [comment]
Looks like OSX failed due to running out of disk.

**Review by sethbrenith — APPROVED** [review]
:shipit:

**Comment by MSLaguana** [comment]
Looks like there is a node test `test-vm-proxy-failure-CP` which explicitly forbids us to call `getOwnPropertyDescriptors` on the sandbox. I'm inclined to just ignore that test for now.

**Comment by MSLaguana** [comment]
Updated CI: https://ci.nodejs.org/job/chakracore-test-pull-request/263/

File: deps/chakrashim/src/jsrtcontextshim.h [hunk 1]
```diff
@@ -98,6 +98,10 @@ class ContextShim {
   void * GetAlignedPointerFromEmbedderData(int index);
   void SetAlignedPointerInEmbedderData(int index, void * value);
 
+  void CacheGlobalProperties();
+  void ResolveGlobalChanges(JsValueRef sandbox);
+
+
```
Comment (kfarnung) [inline deps/chakrashim/src/jsrtcontextshim.h]: nit: extra newline

File: deps/chakrashim/lib/chakra_shim.js [hunk 2]
```diff
@@ -671,6 +672,76 @@
...
+          a.writable !== b.writable) {
+        return false
+      }
+      if (a.get !== b.get || a.set !== b.set) {
+        return false;
+      }
+      return true;
+    }
+    utils.afterContext = function (beforeDescriptors, contextGlobal, sandbox) {
+      try{
```
Comment (kfarnung) [inline deps/chakrashim/lib/chakra_shim.js]: nit: spacing around braces.  I think the linter on Windows should have caught it.

File: deps/chakrashim/src/jsrtcontextshim.h (hunk 1 above)
Comment (MSLaguana) [reply deps/chakrashim/src/jsrtcontextshim.h]: Fixed

File: deps/chakrashim/lib/chakra_shim.js (hunk 2 above)
Comment (MSLaguana) [reply deps/chakrashim/lib/chakra_shim.js]: Fixed.

**Review by kfarnung — APPROVED** [review]

**Pull request merged**